File: Sources/LoeBalance/Refresh/BalanceReconciler.swift
import Foundation

struct BalanceReconciler: Sendable {
    let tolerance = Money(decimal: 0.0001)

    func reconcile(
        previousBalance: Money,
        currentBalance: Money,
        unseenUsage: [UsageRecord]
    ) -> [BalanceAnimationEvent] {
        let sorted = unseenUsage.sorted { ($0.createdAt, $0.id) < ($1.createdAt, $1.id) }
        let debitTotal = sorted.reduce(.zero) { $0 + $1.actualCost }
        var events = debitEvents(from: sorted)
        let residual = (currentBalance - previousBalance) + debitTotal
        if residual > tolerance {
            events.append(.credit(residual))
        } else if residual < Money(decimal: -tolerance.decimal) {
            events.append(.debit(residual.magnitude))
        }
        return events
    }

    func debitEvents(from usage: [UsageRecord]) -> [BalanceAnimationEvent] {
        guard usage.count > 20 else {
            return usage.map { .debit($0.actualCost) }
        }

        let individual = usage.prefix(19).map { BalanceAnimationEvent.debit($0.actualCost) }
        let remainder = usage.dropFirst(19).reduce(.zero) { $0 + $1.actualCost }
        return individual + [.debit(remainder)]
    }
}

File: windows/src/LoeBalance.Core/Refresh/BalanceReconciler.cs
using LoeBalance.Core.Models;

namespace LoeBalance.Core.Refresh;

public sealed class BalanceReconciler
{
    private static readonly Money Tolerance = new(0.0001m);

    public IReadOnlyList<BalanceAnimationEvent> Reconcile(
        Money previousBalance,
        Money currentBalance,
        IEnumerable<UsageRecord> unseenUsage)
    {
        var sorted = unseenUsage.OrderBy(record => record.CreatedAt).ThenBy(record => record.Id).ToArray();
        var debitTotal = sorted.Aggregate(Money.Zero, (total, record) => total + record.ActualCost);
        var events = DebitEvents(sorted).ToList();
        var residual = currentBalance - previousBalance + debitTotal;
        if (residual > Tolerance) events.Add(new BalanceAnimationEvent.Credit(residual));
        else if (residual < new Money(-Tolerance.Decimal)) events.Add(new BalanceAnimationEvent.Debit(residual.Magnitude));
        return events;
    }

    public IReadOnlyList<BalanceAnimationEvent> DebitEvents(IReadOnlyList<UsageRecord> usage)
    {
        if (usage.Count <= 20) return usage.Select(record => new BalanceAnimationEvent.Debit(record.ActualCost)).ToArray();
        var individual = usage.Take(19).Select(record => new BalanceAnimationEvent.Debit(record.ActualCost)).ToList();
        var remainder = usage.Skip(19).Aggregate(Money.Zero, (total, record) => total + record.ActualCost);
        individual.Add(new BalanceAnimationEvent.Debit(remainder));
        return individual;
    }
}
